### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by HongshanLi**
Added validate functions to Trainer class in 'train_loop.py'. This avoids making extra copy of the model.

### CODE REVIEW

File: dataset.py [hunk 1]
```diff
@@ -345,13 +345,8 @@ def test_compute_mean_std(dataset):
...
-        transforms.ToPILImage(),
-        transforms.ToTensor()])
-
-    data_dir = '/home/hongshan/data/Trees_processed/'
-    ds = TreeDataset(data_dir, transform=transform, elv_transform=transform,
-            mask_transform=transform, purpose=None)
-    mean,std = compute_mean_std(ds)
-    print(mean, std)
+
+    data_dir = '/mnt/efs/Trees_processed/'
```
Comment (mrubash1) [inline dataset.py]: Why are you using the EFS instead of S3? That said, I like EFS as you can use it from different instances, so I'm more curious then anything else :)

File: run.sh [hunk 2]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 python main.py /mnt/efs/Trees_processed/ -j=2 --lr=0.0001 -b=16 \
```
Comment (mrubash1) [inline run.sh]: Can you use longer notation for your variable like J lr B, etc?

That way users can more easily use it without have to look up documentation

**Comment by mrubash1** [comment]
-  Can you rename criterion.py to include more information like criterion_loss_function.py
- For your python files darknet.py, detector.py, preprocess.py, and others, did you write all of this code? If not, where did you obtain it? And if you got it from another source, no worries, but could you add a header to each of these files which explains where it is from or modified from?
- Ideas.txt is empty, can you delete it? And/or move that info to the readme instead
- In insepct_data.py, can you either
  - Put all hardcoded information to the top of the file under imports (i.e. annotation, data_dir)
  - Or better, make this a function where the input args include these strings with the current values as defaults
- mean_std is empty, can you delete this file?
- For your class Mask(nn.Module) in model.py, can you not hardcode so many things (i.e. input and output channels, kernel_size, etc). Since there is alot here, at the least just put all the hardcoded variables into the init part of the class (and add a todo into the readme to put this info in a config file in the future if possible)

- Not needed for this PR, but for the first PR in week 4
  - Can you create a conda or pyenv environment that is installable with instructions from the command line? You can use my insight project framework for this which shows a template of how to do it, including readme code

File: dataset.py (hunk 1 above)
Comment (HongshanLi) [reply dataset.py]: Hey Matt, 
Didn't see your reply until now. Yeah I used EFS precisely because of that. I mounted my EFS to one cpu instance and one gpu instance. I do all the coding on the cpu instance. when it is good to 
train I just switch to the gpu instance

File: run.sh (hunk 2 above)
Comment (HongshanLi) [reply run.sh]: Those short options actually have corresponding long options. For example 
`-j = --num-workers`
`-b=--batch-size`

**Comment by HongshanLi** [comment]
The darknet.py comes from another repo, but I ended up using ResNet152 instead of darknet
